### ISSUE TITLE
Fatal Error on reading data when attempting to page results

### ISSUE THREAD

**Issue by ciaran-moore** [MichaelRoosz/php-cassandra#1]
I am trying to page results being returned from queries, but receive either `PHP Fatal error:  Uncaught Cassandra\Response\Exception: Tried to read more data than available in cassandra_test/vendor/mroosz/php-cassandra/src/Response/StreamReader.php:482` or I am getting the same set of results when setting the `page_state` option based on the type of request.

Cassandra Version: `5.0-BETA1`
OS: `Mac Sonoma 14.0 (23A344) - Apple M1 Pro`
PHP Version: `8.3.6`

### Create Keyspace / Table:
```cql
CREATE KEYSPACE test WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;

CREATE TABLE test.storage (
    filename text,
    ukey text,
    value map<text, text>,
    PRIMARY KEY (filename, ukey)
) WITH CLUSTERING ORDER BY (ukey ASC);
```

### Generate Data
```php
$nodes = [];
foreach($config['hosts'] as $h)
{
    $nodes[] =[             // advanced way, using Connection\Stream, persistent connection
            'host'      => $h,
            'port'      => $config['port'],
            'username'  => $config['username'],
            'password'  => $config['password'],
            // 'class'     => 'Cassandra\Connection\Stream',//use stream instead of socket, default socket. Stream may not work in some environment
            'socket'      => [SO_RCVTIMEO => ["sec" => 10, "usec" => 0]], //socket transport only
            'connectTimeout' => 10, // connection timeout, default 5,  stream transport only
            'timeout'   => 120, // write/recv timeout, default 30, stream transport only
            'persistent'    => true, // use persistent PHP connection, default false,  stream transport only  
    ];
}

            $connection = new Cassandra\Connection($nodes, $config['keyspace']);
            $connection->setConsistency(Cassandra\Request\Request::CONSISTENCY_ONE);

            $batches = [];
            $testItem = json_decode('{
                        "id": "4003578",
                        "image": "https://example.music-retailer.com/media-dynamic/images/product/music/album/image0/christmas_by_the_bay-1001178-frntl.jpg",
                        "description": "null",
                        "title": "Christmas By The River (CD)",
                        "taxonomy": "Music > Pop > 80\'s artists"
                        }', true);


            $batch = new Cassandra\Request\Batch(Cassandra\Request\Batch::TYPE_UNLOGGED, Cassandra\Request\Request::CONSISTENCY_ONE);

            $counter = $i = $id = $total_counter = 0;
            $partition= 1;
            $start = microtime(true);

            while (true)
            {
                $i++;
                $total_counter++;
                $counter++;
                $id++;

                $batch->appendQuery(
                    'INSERT INTO storage(filename, ukey, value) VALUES(:filename,:ukey,:value)',
                    [
                        new Cassandra\Type\Varchar('p7_'.$partition),
                        new Cassandra\Type\Varchar('key_'.$id),
                        new Cassandra\Type\CollectionMap($testItem, [
                            // [
                            Cassandra\Type::VARCHAR,
                            Cassandra\Type::VARCHAR,
                            Cassandra\Type::VARCHAR,
                            Cassandra\Type::VARCHAR,
                            Cassandra\Type::VARCHAR
                            // ]
                        ])
                    ]                        
                );

                if (microtime(true) - $start > 5)
                {
                    $taken = round(microtime(true) - $start,2);
                    $start = microtime(true);

                    echo "taken: $taken | items: $counter ( total: $i) | partition: $partition | PHP Ops/sec: " .
                        round($total_counter / $taken) ." | "."memory used : ".
                        round((memory_get_usage()/1024)/1024)."MB\n";
                    $counter = 0;
                }

                if($counter > 200)
                {
                    $batches[] = $batch;
                    $batch = new Cassandra\Request\Batch(Cassandra\Request\Batch::TYPE_UNLOGGED, Cassandra\Request\Request::CONSISTENCY_ONE);
                    $counter = 0;
                }

                if($id >= 100000)
                {
                    $partition++;
                    $id = 0;
                }


                if ($i >= 200000) break;
            }

            echo count($batches) . "\n";
            $start = microtime(true);
            foreach($batches as $batch)
            {
                usleep(500);
                $response = $connection->asyncRequest($batch);
            }

            $taken = round(microtime(true) - $start,2);
            echo "taken: $taken | items: $counter ( total: $i) | partition: $partition | Cass Ops/sec: " .
                round($total_counter / $taken) ." | "."memory used : ".
                round((memory_get_usage()/1024)/1024)."MB\n";
```

#### Fatal error: Tried to read more data than available
```php
$preparedData = $connection->prepare("SELECT * FROM storage where filename = 'p7_2'");
$strictValues = \Cassandra\Request\Request::strictTypeValues(
    [],
    $preparedData['metadata']['columns']
);

$result = $connection->executeSync(
    'p7_2',
    $strictValues,
    \Cassandra\Request\Request::CONSISTENCY_QUORUM,
    [
        'page_size' => 100,
        'names_for_values' => true,
        'skip_metadata' => true,
    ]
);

// Fails at $connection->prepare()
```

#### Same result set even though `page_state` is specified.
```php
    $query = "SELECT * FROM storage where filename = 'p7_2'";
    $statement1 = $connection->querySync($query, 
        [],
        Cassandra\Request\Request::CONSISTENCY_ONE,
        [
            'page_size' => 100,
        ]
    );

    $state = $statement1->getMetadata()['page_state'] ?? false;
    echo json_encode($statement1->fetchRow()) . "\n" .$state ."\n\n";

 // == Q2
    $statement1 = $connection->querySync($query, 
        [],
        Cassandra\Request\Request::CONSISTENCY_ONE,
        [
            'page_size' => 100,
            'page_state' => $state
        ]
    );    

    $state = $statement1->getMetadata()['page_state'] ?? false;
    echo json_encode($statement1->fetchRow()) . "\n" .$state ."\n\n";


```

**Comment by MichaelRoosz** [comment]
Hello @ciaran-moore ,
thank you for opening this issue.

- the prepared statement was failing due to a bug in the library and should be fixed with v0.8.0.
- the phpdoc documentation for "page_state" was wrong - it should be "paging_state"

After slightly adjusting your test code for the updated prepared statement api in v0.8.0 it seems to work now:

test 1
```php
$preparedData = $connection->prepare('SELECT * FROM storage where filename = :filename');
$result = $connection->executeSync(
    $preparedData,
    [
        'filename' => 'p7_2',
    ],
    \Cassandra\Request\Request::CONSISTENCY_QUORUM,
    [
        'page_size' => 100,
        'names_for_values' => true,
    ]
);

$state = $result->getMetadata()['paging_state'] ?? false;
echo json_encode($result->fetchRow()) . "\n" . $state . "\n\n";

// == Q2
$result = $connection->executeSync(
    $result,
    [
        'filename' => 'p7_2',
    ],
    \Cassandra\Request\Request::CONSISTENCY_QUORUM,
    [
        'page_size' => 100,
        'names_for_values' => true,
        'paging_state' => $state,
    ]
);

$state = $result->getMetadata()['paging_state'] ?? false;
echo json_encode($result->fetchRow()) . "\n" . $state . "\n\n";
```

test 2
```php
$query = "SELECT * FROM storage where filename = 'p7_2'";
$statement1 = $connection->querySync($query,
    [],
    Cassandra\Request\Request::CONSISTENCY_ONE,
    [
        'page_size' => 100,
    ]
);

// == Q2
$state = $statement1->getMetadata()['paging_state'] ?? false;
echo json_encode($statement1->fetchRow()) . "\n" . $state . "\n\n";

$statement1 = $connection->querySync($query,
    [],
    Cassandra\Request\Request::CONSISTENCY_ONE,
    [
        'page_size' => 100,
        'paging_state' => $state,
    ]
);

$state = $statement1->getMetadata()['paging_state'] ?? false;
echo json_encode($statement1->fetchRow()) . "\n" . $state . "\n\n";
```

output
```
taken: 5 | items: 15 ( total: 138303) | partition: 2 | PHP Ops/sec: 27661 | memory used : 66MB
994
taken: 1.06 | items: 191 ( total: 200000) | partition: 3 | Cass Ops/sec: 188679 | memory used : 96MB
{"filename":"p7_2","ukey":"key_100","value":{"description":"null","id":"4003578","image":"https:\/\/example.music-retailer.com\/media-dynamic\/images\/product\/music\/album\/image0\/christmas_by_the_bay-1001178-frntl.jpg","taxonomy":"Music > Pop > 80's artists","title":"Christmas By The River (CD)"}}

	key_10089????????

{"filename":"p7_2","ukey":"key_1009","value":{"description":"null","id":"4003578","image":"https:\/\/example.music-retailer.com\/media-dynamic\/images\/product\/music\/album\/image0\/christmas_by_the_bay-1001178-frntl.jpg","taxonomy":"Music > Pop > 80's artists","title":"Christmas By The River (CD)"}}

	key_10179???7???7

{"filename":"p7_2","ukey":"key_100","value":{"description":"null","id":"4003578","image":"https:\/\/example.music-retailer.com\/media-dynamic\/images\/product\/music\/album\/image0\/christmas_by_the_bay-1001178-frntl.jpg","taxonomy":"Music > Pop > 80's artists","title":"Christmas By The River (CD)"}}

	key_10089????????

{"filename":"p7_2","ukey":"key_1009","value":{"description":"null","id":"4003578","image":"https:\/\/example.music-retailer.com\/media-dynamic\/images\/product\/music\/album\/image0\/christmas_by_the_bay-1001178-frntl.jpg","taxonomy":"Music > Pop > 80's artists","title":"Christmas By The River (CD)"}}

	key_10179???7???7
```

**Comment by ciaran-moore** [comment]
Fantastic Michael. I have done a pull and tested now, and can confirm this has resolved the issue.

```
php read_test.php
// single item
{"filename":"p7_2","ukey":"key_1","value":{"description":"null","id":"4003578","image":"https:\/\/example.music-retailer.com\/media-dynamic\/images\/product\/music\/album\/image0\/christmas_by_the_bay-1001178-frntl.jpg","taxonomy":"Music > Pop > 80's artists","title":"Christmas By The River (CD)"}}

	key_10086????????
// loop all results and show count
{"filename":"p7_2","ukey":"key_10087","value":{"description":"null","id":"4003578","image":"https:\/\/example.music-retailer.com\/media-dynamic\/images\/product\/music\/album\/image0\/christmas_by_the_bay-1001178-frntl.jpg","taxonomy":"Music > Pop > 80's artists","title":"Christmas By The River (CD)"}}

	key_10176???7???7

100100 Items
```
